diff --color -ruN gcc-4.9.4/gcc/config/sh/sh-c.c gcc-4.9.4-kos/gcc/config/sh/sh-c.c
--- gcc-4.9.4/gcc/config/sh/sh-c.c	2023-05-17 10:54:39.317087863 -0500
+++ gcc-4.9.4-kos/gcc/config/sh/sh-c.c	2023-05-17 10:55:02.502142590 -0500
@@ -145,4 +145,11 @@
 
   cpp_define_formatted (pfile, "__SH_ATOMIC_MODEL_%s__",
 			selected_atomic_model ().cdef_name);
+
+  /* Custom built-in defines for KallistiOS */
+  builtin_define ("__KOS_GCC_PATCHED__");
+  cpp_define_formatted (pfile, "__KOS_GCC_PATCHLEVEL__=%d",
+			2023010200);
+  /* Toolchain supports setting up stack for 32MB */
+  builtin_define ("__KOS_GCC_32MB__");
 }
diff --color -ruN gcc-4.9.4/gcc/configure gcc-4.9.4-kos/gcc/configure
--- gcc-4.9.4/gcc/configure	2023-05-17 10:54:38.019084799 -0500
+++ gcc-4.9.4-kos/gcc/configure	2023-05-17 10:55:02.504142595 -0500
@@ -11214,7 +11214,7 @@
     target_thread_file='single'
     ;;
   aix | dce | lynx | mipssde | posix | rtems | \
-  single | tpf | vxworks | win32)
+  single | tpf | vxworks | win32 | kos)
     target_thread_file=${enable_threads}
     ;;
   *)
diff --color -ruN gcc-4.9.4/gcc/cp/cfns.h gcc-4.9.4-kos/gcc/cp/cfns.h
--- gcc-4.9.4/gcc/cp/cfns.h	2023-05-17 10:54:39.281087778 -0500
+++ gcc-4.9.4-kos/gcc/cp/cfns.h	2023-05-17 10:55:02.504142595 -0500
@@ -89,7 +89,7 @@
       400, 400, 400, 400, 400, 400, 400, 400, 400, 400,
       400, 400, 400, 400, 400, 400, 400
     };
-  register int hval = len;
+  register int hval = (int)len;
 
   switch (hval)
     {

diff --color -ruN gcc-4.9.4/libgcc/config/sh/t-sh gcc-4.9.4-kos/libgcc/config/sh/t-sh
--- gcc-4.9.4/libgcc/config/sh/t-sh	2023-05-17 10:54:37.857084416 -0500
+++ gcc-4.9.4-kos/libgcc/config/sh/t-sh	2023-05-17 10:55:02.504142595 -0500
@@ -23,6 +23,8 @@
   $(LIB1ASMFUNCS_CACHE)
 LIB1ASMFUNCS_CACHE = _ic_invalidate _ic_invalidate_array
 
+LIB2ADD = $(srcdir)/config/sh/fake-kos.S
+
 crt1.o: $(srcdir)/config/sh/crt1.S
 	$(gcc_compile) -c $<
 
diff --color -ruN gcc-4.9.4/libgcc/configure gcc-4.9.4-kos/libgcc/configure
--- gcc-4.9.4/libgcc/configure	2023-05-17 10:54:37.835084364 -0500
+++ gcc-4.9.4-kos/libgcc/configure	2023-05-17 10:55:02.505142597 -0500
@@ -4571,6 +4571,7 @@
     tpf)	thread_header=config/s390/gthr-tpf.h ;;
     vxworks)	thread_header=config/gthr-vxworks.h ;;
     win32)	thread_header=config/i386/gthr-win32.h ;;
+    kos)	thread_header=config/sh/gthr-kos.h ;;
 esac
 
 
diff --color -ruN gcc-4.9.4/libstdc++-v3/config/cpu/sh/atomicity.h gcc-4.9.4-kos/libstdc++-v3/config/cpu/sh/atomicity.h
--- gcc-4.9.4/libstdc++-v3/config/cpu/sh/atomicity.h	2023-05-17 10:54:37.720084093 -0500
+++ gcc-4.9.4-kos/libstdc++-v3/config/cpu/sh/atomicity.h	2023-05-17 10:55:02.505142597 -0500
@@ -80,6 +80,12 @@
 namespace 
 {
   __gnu_cxx::__mutex atomic_mutex;
+  __gnu_cxx::__mutex&
+  get_atomic_mutex()
+  {
+    static __gnu_cxx::__mutex atomic_mutex;
+    return atomic_mutex;
+  }
 } // anonymous namespace
 
 namespace __gnu_cxx _GLIBCXX_VISIBILITY(default)
diff --color -ruN gcc-4.9.4/libstdc++-v3/configure gcc-4.9.4-kos/libstdc++-v3/configure
--- gcc-4.9.4/libstdc++-v3/configure	2023-05-17 10:54:37.386083305 -0500
+++ gcc-4.9.4-kos/libstdc++-v3/configure	2023-05-17 10:55:02.509142606 -0500
@@ -15193,6 +15193,7 @@
     tpf)	thread_header=config/s390/gthr-tpf.h ;;
     vxworks)	thread_header=config/gthr-vxworks.h ;;
     win32)	thread_header=config/i386/gthr-win32.h ;;
+    kos)    thread_header=config/sh/gthr-kos.h ;;
 esac
 
 
